File: cpp/probe_test/probe_test.cpp
// probe_test -- C++ port of examples\Vb6\probe_test\probe_test.bas
// Step-by-step probe of the TPDF flow (in-memory CreateNewPDF + OpenOutputFile + TPDFTable).
#include <lumaspdf.h>
#include <cstdio>
#include <string>

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    printf("ERR %d: %s\n", ErrCode, ErrMessage ? ErrMessage : "");
    return 0;
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    printf("Create ok\n");
    pdfSetOnErrorProc(pdf, nullptr, ErrProc);
    printf("CreateNewPDF('') = %d\n", pdfCreateNewPDFA(pdf, ""));
    printf("SetPageCoords = %d\n", pdfSetPageCoords(pdf, pcTopDown));
    printf("Append = %d\n", pdfAppend(pdf));
    printf("SetFont = %d\n", pdfSetFontA(pdf, "Arial", fsRegular, 12.0, 1, cp1252));
    printf("WriteText = %d\n", pdfWriteTextA(pdf, 50, 50, "probe"));

    ITBL tbl = tblCreateTable(pdf, 3, 3, 500.0f, 100.0f);
    printf("Table created\n");
    SI32 r = tblAddRow(tbl, -1.0f);
    printf("AddRow = %d\n", r);
    printf("SetCellText = %d\n", tblSetCellTextA(tbl, r, 0, taLeft, coTop, "cell", (UI32)-1));
    printf("DrawTable = %f\n", tblDrawTable(tbl, 50.0f, 80.0f, 700.0f));
    printf("HaveMore = %d\n", tblHaveMore(tbl));
    tblDeleteTable(&tbl);

    printf("EndPage = %d\n", pdfEndPage(pdf));
    printf("GetPageCount = %d\n", pdfGetPageCount(pdf));
    printf("HaveOpenDoc = %d\n", pdfHaveOpenDoc(pdf));
    std::string outFile = exeDir(argv[0]) + "/probe_out.pdf";
    printf("OpenOutputFile = %d\n", pdfOpenOutputFileA(pdf, outFile.c_str()));
    printf("CloseFile = %d\n", pdfCloseFile(pdf));
    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/probe_test/probe_test.cs
// probe_test -- C# port of examples\Vb6\probe_test\probe_test.bas
// Step-by-step probe of the TPDF flow + TPDFTable (tbl* exports).
using System;
using System.IO;
using LumasPdfSdk;

class ProbeTest
{
    static TErrorProc _err = ErrProc;

    static int ErrProc(IntPtr data, int errCode, string errMessage, int errType)
    {
        Console.WriteLine("ERR " + errCode + ": " + errMessage);
        return 0;
    }

    static void Main()
    {
        string dir = AppDomain.CurrentDomain.BaseDirectory;
        IntPtr pdf = LumasPdf.pdfNewPDF();
        Console.WriteLine("Create ok");
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        Console.WriteLine("CreateNewPDF('') = " + LumasPdf.pdfCreateNewPDFW(pdf, ""));
        Console.WriteLine("SetPageCoords = " + LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown));
        Console.WriteLine("Append = " + LumasPdf.pdfAppend(pdf));
        Console.WriteLine("SetFont = " + LumasPdf.pdfSetFontW(pdf, "Arial", LumasPdfConsts.fsRegular, 12.0, true, TCodepage.cp1252));
        Console.WriteLine("WriteText = " + LumasPdf.pdfWriteTextW(pdf, 50, 50, "probe"));

        IntPtr tbl = LumasPdf.tblCreateTable(pdf, 3, 3, 500f, 100f);
        Console.WriteLine("Table created");
        int r = LumasPdf.tblAddRow(tbl, -1f);
        Console.WriteLine("AddRow = " + r);
        // NOTE: the VB6 template passes Len = -1 (the classic API's null-terminated sentinel), but the
        // LumasPdf engine treats Len as an unsigned byte count (Len>0 => copy Len bytes), so
        // 0xFFFFFFFF causes a 4GB copy / AccessViolation. Pass 0 = auto-length instead.
        Console.WriteLine("SetCellText = " + LumasPdf.tblSetCellTextA(tbl, (uint)r, 0, LumasPdfConsts.taLeft, TCellAlign.coTop, "cell", 0));
        Console.WriteLine("DrawTable = " + LumasPdf.tblDrawTable(tbl, 50f, 80f, 700f));
        Console.WriteLine("HaveMore = " + LumasPdf.tblHaveMore(tbl));
        LumasPdf.tblDeleteTable(ref tbl);

        Console.WriteLine("EndPage = " + LumasPdf.pdfEndPage(pdf));
        Console.WriteLine("GetPageCount = " + LumasPdf.pdfGetPageCount(pdf));
        Console.WriteLine("HaveOpenDoc = " + LumasPdf.pdfHaveOpenDoc(pdf));
        string outFile = Path.Combine(dir, "probe_out.pdf");
        Console.WriteLine("OpenOutputFile = " + LumasPdf.pdfOpenOutputFileW(pdf, outFile));
        Console.WriteLine("CloseFile = " + LumasPdf.pdfCloseFile(pdf));
        LumasPdf.pdfDeletePDF(pdf);
    }
}
